pennylane/data/base/dataset.py

```diff
@@ -0,0 +1,419 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the :class:`~pennylane.data.Dataset` base class, and `qml.data.Attribute` class
+for declaratively defining dataset classes.
+"""
+
+import typing
+from dataclasses import InitVar, dataclass
+from pathlib import Path
+from types import MappingProxyType
+from typing import (
+    Any,
+    ClassVar,
+    Generic,
+    List,
+    Literal,
+    Optional,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+    get_origin,
+    Tuple,
+)
+
+# pylint doesn't think this exists
+from typing_extensions import dataclass_transform  # pylint: disable=no-name-in-module
+
+from pennylane.data.base import hdf5
+from pennylane.data.base.attribute import AttributeInfo, DatasetAttribute
+from pennylane.data.base.hdf5 import HDF5Any, HDF5Group, h5py
+from pennylane.data.base.mapper import MapperMixin, match_obj_type
+from pennylane.data.base.typing_util import UNSET, T
+
+
+@dataclass
+class Field(Generic[T]):
+    """
+    The Attribute class is used to declaratively define the
+    attributes of a Dataset subclass, in a similar way to
+    dataclasses. This class should not be used directly when
+    declaring attributes, use the ``attribute()`` function instead.
```

Comment:
wondering if this is a copy-paste or if I'm not following - should a lot of these instances of the word "attribute" instead be "field"?

Reply:
Yeah, my bad.